Add Part 2 Chapter 3 (#33)
ISSOtm/gb-asm-tutorial: src/part2/functions.md

# Functions

Earlier in this tutorial you copied various graphics from ROM into VRAM.
This operation is conventionally known as `Memcpy`, and by making it a function we can reuse the code in many places.
Write this below the `UpdateKeys` function:

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/functions/main.asm:memcpy}}
{{#include ../../unbricked/functions/main.asm:memcpy}}
```

> In languages like C, functions automatically return when they reach the end of their scope.
> However, functions in assembly don't have a definitive "end"; you always need to manually place a `ret` instruction at the end of the function to return from it

Reviewer: ```suggestion
The new `ret` instruction should immediately catch our eye.
It is, unsurprisingly, what makes execution *return* to where the function was *called* from.
Importantly, many languages have a definite "end" to a function: in C or Rust, that's the closing brace `}`; in Pascal or Lua, the keyword `end`, and so on; the function implicitly returns when execution reaches its end.
However, **this is not the case in assembly**, so you must remember to add a `ret` instruction at the end of the function to return from it!
Otherwise, the results are unpredictable.
```

> Notice that the function has a comment explaining which registers it takes as input.

Reviewer: ```suggestion
Notice the comment above the function, explaining which registers it takes as input.
```

> This is important so that you know how to interface with the function.

Reviewer: ```suggestion
This is important so that you know how to interface with the function: assembly has no formal parameters, so comments explaining them are even more important than with other languages.
```

> We'll see more of this later on.

Reviewer: What is "this" here?

Author: The comment explaining the function's interface

Author: I've updated the wording very slightly; this might be better now

Reviewer: ```suggestion
We'll see more of those as we progress.
```
I think this conveys the intent better?

> There are three places in your initialization code where you can use the `Memcpy` functions.

Reviewer: ```suggestion
There are three places in the initialization code where we can use the `Memcpy` function.
```

Find each of these copy loops and replace them with a call to `Memcpy`.

> Make sure to leave the registers as-is; these are the parameters to the function.

Reviewer: ```suggestion
The registers serve as parameters to the function, so we'll leave them as-is.
```

# Functions

Earlier in this tutorial you copied various graphics from ROM into VRAM.
This operation is conventionally known as `Memcpy`, and by making it a function we can reuse the code in many places.
Write this below the `UpdateKeys` function:

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/input/main.asm:memcpy}}
{{#include ../../unbricked/input/main.asm:memcpy}}

> ```

Reviewer: The "Input" lesson is after this one; can you make an intermediate copy of the code as well, please?

In languages like C, functions automatically return when they reach the end of their scope.

> However, functions in assembly don't have a definative "end"; you always need to manually place a `ret` instruction at the end of the function to return from it.

Reviewer 2: ```suggestion
However, functions in assembly don't have a definitive "end"; you always need to manually place a `ret` instruction at the end of the function to return from it.
```

> The mechanics behind `call` and `ret` can be a bit complicated, so we'll explain how they work later.

Reviewer 2: Why not introduce them in a "simplistic" way first then dive in later?

Reviewer 3: I don't see how it could be introduced at this point in the tutorial.

Author: Yeah, I don't see any way to give a satisfying explanation without also explaining the whole stack. Saying "it's like `return` in C" is find and understandable and won't cause any issues until `push`/`pop` are introduced.

Reviewer 3: remind me again what push and pop do?

Reviewer: I agree with avivace insofar that we should present a simple behaviour model for now, while leaving in a tip box that "the actual behaviour is much more subtle, but we will explore it later".

Reviewer 2: @ISSOtm can you suggest a way we could introduce the call and ret functions which not requires the full explanation? IMO I don't see how introducing them in a very simple way can be blocking, e.g. : http://www.c-jump.com/CIS77/ASM/Procedures/P77_0010_call_ret.htm

Reviewer: Is this good enough? https://github.com/ISSOtm/gb-asm-tutorial/pull/33/files#r1038784843

Reviewer 2: @ISSOtm that link is simply pointing to the pr diff

Reviewer: The link is actually pointing to the comment whose suggestion you just accepted, but GitHub doesn't show it in the diff view since it's "Outdated". Here is a link to the same in the "Conversation" tab, which seems to work: https://github.com/ISSOtm/gb-asm-tutorial/pull/33#discussion_r1038784843

Reviewer 2: Oh okay, then! Parfait

> Notice that the function has a comment explaining which registers it takes as input.

Reviewer 3: If this is important to know, why not explain it a little now, if not just the basics of it? it makes no sense to me that we're just putting off an important concept related to input and functions, especially since it doesn't look like it would take long to explain.

Reviewer: Because the full explanation is really complicated, and just building the game is already difficult enough as-is.

Reviewer 3: alrighty then.

<div class="table-wrapper"><table><thead><tr><th>Before</th><th>After</th></tr></thead><tbody><tr><td>

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:copy_tiles}}
{{#include ../../unbricked/getting-started/main.asm:copy_tiles}}
```

</td><td>

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/functions/main.asm:copy_tiles}}
{{#include ../../unbricked/functions/main.asm:copy_tiles}}
```

</td></tr><tr><td>

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:copy_map}}
{{#include ../../unbricked/getting-started/main.asm:copy_map}}
```

</td><td>

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/functions/main.asm:copy_map}}
{{#include ../../unbricked/functions/main.asm:copy_map}}
```

</td></tr><tr><td>

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/objects/main.asm:copy-paddle}}
{{#include ../../unbricked/objects/main.asm:copy-paddle}}
```

</td><td>

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/functions/main.asm:copy_paddle}}
{{#include ../../unbricked/functions/main.asm:copy_paddle}}
```

</td></tr></tbody></table></div>

> In the next chapter we'll write a second function to read player input.

Reviewer: ```suggestion
In the next chapter, we'll write another function, this time to read player input.
```

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/functions/main.asm:memcpy}}
{{#include ../../unbricked/functions/main.asm:memcpy}}
```

The new `ret` instruction should immediately catch our eye.
It is, unsurprisingly, what makes execution *return* to where the function was *called* from.
Importantly, many languages have a definite "end" to a function: in C or Rust, that's the closing brace `}`; in Pascal or Lua, the keyword `end`, and so on; the function implicitly returns when execution reaches its end.
However, **this is not the case in assembly**, so you must remember to add a `ret` instruction at the end of the function to return from it!
Otherwise, the results are unpredictable.

Notice the comment above the function, explaining which registers it takes as input.
This is important so that you know how to interface with the function: assembly has no formal parameters, so comments explaining them are even more important than with other languages.
We'll see more of this later on.

There are three places in the initialization code where we can use the `Memcpy` function.

> Find each of these copy loops and replace them with a call to `Memcpy`.

Reviewer: ```suggestion
Find each of these copy loops and replace them with a call to `Memcpy`; for this, we use the `call` instruction.
```
